mitiq/pec/pec.py

```diff
@@ -0,0 +1,115 @@
+        executor = A function which executes a circuit and returns an
+            expectation value.
+        deco_dict = The decomposition dictionary containing the quasi-
+            probability representation of the ideal operations (those
+            which are part of the input circuit).
+        num_to_average: The number times the expectation value of each sampled
+            circuit is evaluated, then averaged. Setting this number to values
+            larger than 1 (default), reduces the statistical error associated
+            to each noisy expectation value.
+        num_samples: The number of noisy circuits to be sampled for PEC.
+            If equal to None, it is deduced from the amount of "negativity"
+            of the quasi-probability representation of the input circuit.
+            Note: the latter feature is not yet implemented and num_samples
+            is just set to 1000 if not specified.
+
+    Returns:
+        The PEC estimate of the ideal expectation value associated
+        to the input circuit.
+
+    .. [Temme2017] : Kristan Temme, Sergey Bravyi, Jay M. Gambetta,
+        "Error mitigation for short-depth quantum circuits,"
+        *Phys. Rev. Lett.* **119**, 180509 (2017),
+        (https://arxiv.org/abs/1612.02058).
+
+    .. [Endo2018] : Suguru Endo, Simon C. Benjamin, Ying Li,
+        "Practical Quantum Error Mitigation for Near-Future Applications"
+        *Phys. Rev. **X 8**, 031027 (2018),
+        (https://arxiv.org/abs/1712.09271).
+
+    .. [Takagi2020] : Ryuji Takagi,
+        "Optimal resource cost for error mitigation,"
+        (https://arxiv.org/abs/2006.12509).
+    """
+
+    # TODO: maybe deduce num_to_sample from the decomposition "negativity"
+    if not num_samples:
+        num_samples = 1000
+
+    sampled_circuits = []
+    signs = []
+    norms = []
+
+    for _ in range(num_samples):
+        sampled_circuit, sign, norm = sample_circuit(circuit, deco_dict)
+        sampled_circuits.append(sampled_circuit)
+        signs.append(sign)
+        norms.append(norm)
+
+    # The norm of the quasi-distribution should be independent of sample
+    assert np.all(norms == norms[0])
+    norm = norms[0]
+
+    # Repeat each sampled circuit "num_to_average" times
+    to_run = [circ for circ in sampled_circuits for _ in range(num_to_average)]
+
+    # TODO: think about dealing with batched executors
+    # Execute all the circuits
+    results = np.array([executor(circ) for circ in to_run])
+
+    # Reshape results to "num_to_average" columns
+    results = results.reshape(-1, num_to_average)
+
+    # Average the "num_to_average" columns
+    exp_values = np.average(results, axis=1)
+
+    # Evaluate unbiased estimators [Temme2017], [Endo2018], [Takagi2020]
+    unbiased_estimators = [s * norm * val for val, s in zip(exp_values, signs)]
+
+    # Average to return the PEC estimate of the ideal expectation value
+    return np.average(unbiased_estimators)
```

Comment:
Return the variance as well?

Reply:
I added an `full_output` option for returning the standard deviation of the sampled pec values.  I used the standard dev. instead of the variance just for symmetry with respect to ZNE factories.

Note also that, as written in the docstring, this is the "width" of the sampling distribution. To estimate the error between the `pec_value` and the ideal expectation value one should
divide the standard deviation by the square root of the `num_samples`.

I am not sure if this is clear enough for the user.

Reply:
I guess it depends on an understanding of PEC. I would divide it by sqrt(num_samples) and return that since this is the quantity of interest in most cases.

Reply:
Agree. I replaced the previous `pec_std` with `pec_error` which is evaluated as:
`pec_error = np.std(unbiased_estimators) / np.sqrt(num_samples)`